add token auth abilities

jessereynolds opened the pull request.

jessereynolds committed:
add token auth abilities

briancain commented on lib/puppet_https.rb:
```diff
@@ -76,10 +79,27 @@ def post(url, request_body=nil)
+
+  def token
+    return @token if @token
+    token_path = @settings['token_path']
```
Do you think it makes sense to have a sensible default token path to read in? I don't think the path is configurable in PE, so I think it'll always be at `HOMEDIR/.puppetlabs/token`...I think?

jessereynolds replied:
Yeah, that's a good idea. I'll do that.

briancain commented on lib/puppet_https.rb:
```diff
@@ -20,19 +23,19 @@ def make_ssl_request(url, req)
     pkey_path = @settings['private_key_path']
 
-    if File.exists?(ca_file)
+    if ca_file and File.exists?(ca_file)
```
Hmm, is this change to prevent an exception being thrown if the file paths weren't provided (i.e. it tries to do `File.exists?(nil)`)? If so I think this if statement will still throw an exception:

```ruby
[4] pry(main)> if "hello" and File.exists?(nil)
[4] pry(main)*   puts "nope"
[4] pry(main)* end
(pry):6: warning: string literal in condition
TypeError: no implicit conversion of nil into String
from (pry):4:in `exists?'
```

It might be better to set them to empty strings above as a default like so:

```ruby
ca_file = @settings['ca_certificate_path'] ||= ""
```

Then `File.exists?` will evaluate an empty string rather than nil.

jessereynolds replied:
Yes, this is to prevent an exception being thrown if the file paths were not provided. Because ca_file is nil if it's not being defined, the File.exists? will not be evaluated in that case: 

```
irb(main):001:0> ca_file = nil
=> nil
irb(main):002:0> if ca_file and File.exists?(ca_file)
irb(main):003:1> puts "nope"
irb(main):004:1> end
=> nil
```

briancain replied:
Ah ok I see now, thanks for the clarification!

briancain commented:
Thanks so much for the PR! Apologies for the delay. I left a few comments. Haven't tested it yet, but I set some time aside on Friday to do so.

briancain commented:
/cc @beergeek in case he has any additional feedback

briancain commented:
What happens if a user supplies both all of the certs and keys and also a token file? From what I can see it won't use the token and instead will use the certs/keys. Is that expected behavior?

briancain reviewed (changes requested):
Overall looks good and thanks for the PR! I have a few comments, but otherwise 👍

jessereynolds commented:
At the moment I think it's not clear which auth method will actually be used if both token and cert are supplied. Perhaps the PuppetHttps module should be clearer about this. That is:
- specifically prefer one method over the other (in the case that both methods are provided), and
- raise an exception if no auth methods are provided
Thoughts @briancain ?

briancain commented:
@jessereynolds - that sounds perfect to me. I'd probably default to token if both methods are provided personally, and 👍  if no authentication methods were provided.

briancain commented:
Lemme know when you update the pull request and I'll be happy to review again, thanks!

jessereynolds committed:
reorganised a fair bit:
- moved token vs cert logic to initialize
- prefers token auth
- fails early if no auth method is available

jessereynolds committed:
fixed one test

jessereynolds committed:
fix errors, improve testability

jessereynolds committed:
fix and enhance puppet_https specs

jessereynolds committed:
fix all the tests

jessereynolds committed:
Now verifies token file exists and is non empty

jessereynolds commented:
Updated my branch after feedback from @beergeek - we now verify that a token file we're relying on exists and is non empty during initialization, also that a token string passed in is not empty.

jessereynolds committed:
add mention of default token path use to readme

beergeek commented:
I am giving a large 👍

jessereynolds commented:
@briancain Could you give this another review? I've addressed the things we discussed above, and a few other things, namely:
- token auth is now the default (in the case where cert auth and explicit token (or token_path) are supplied)
- default token path (`~/.puppetlabs/token`) is now consulted in the case where no explicit auth parameters are passed in
- moved all the logic around determining authentication method and creds into the initialize method
- raise appropriate exceptions when:
  - token file does not exist, or is empty
  - token is empty
  - specified certificate or key file do not exist
  - no authentication methods are available
- updated spec tests
- added new spec tests

jonathannewman reviewed (approved):
👍
